File: 04-String-Manipulation/1456. Maximum Number of Vowels in a Substring of Given Length/1456. Maximum Number of Vowels in a Substring of Given Length.java
1class Solution {
2
3    boolean solve(char ch) {
4        if (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u') {
5            return true;
6        }
7
8        return false;
9    }
10
11    public int maxVowels(String s, int k) {
12        int n = s.length();
13        int cnt = 0, maxi = 0;
14
15        for (int i = 0; i < k; i++) {
16            if (solve(s.charAt(i)))
17                cnt++;
18        }
19
20        maxi = cnt;
21
22        for (int i = k; i < n; i++) {
23            if (solve(s.charAt(i)))
24                cnt++; // right enters
25            if (solve(s.charAt(i - k)))
26                cnt--; // left leaves
27            maxi = Math.max(maxi, cnt);
28        }
29
30        return maxi;
31    }
32}

File: 04-String-Manipulation/1456. Maximum Number of Vowels in a Substring of Given Length/1456. Maximum Number of Vowels in a Substring of Given Length.cpp
1class Solution {
2public:
3    bool isVowel(char c) {
4        c = tolower(c);
5        return (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u');
6    }
7
8    int maxVowels(string s, int k) {
9        int n = s.size();
10        int c = 0;
11        int maxCount = 0;
12
13        for (int i = 0; i < k; i++) {
14            if (isVowel(s[i]))
15                c++;
16        }
17        maxCount = c;
18
19        for (int i = k; i < n; i++) {
20            if (isVowel(s[i]))
21                c++;
22            if (isVowel(s[i - k]))
23                c--;
24            maxCount = max(maxCount, c);
25        }
26
27        return maxCount;
28    }
29};
30